packages/store/addon/-private/identifiers/cache.ts

```diff
@@ -0,0 +1,426 @@
+
+type IdentifierMap = Dict<string, StableRecordIdentifier>;
+type TypeMap = Dict<string, KeyOptions>;
+
+interface KeyOptions {
+  lid: IdentifierMap;
+  id: IdentifierMap;
+  _allIdentifiers: StableRecordIdentifier[];
+}
+
+let configuredForgetMethod: ForgetMethod;
+let configuredGenerationMethod: GenerationMethod;
+let configuredResetMethod: ResetMethod;
+let configuredUpdateMethod: UpdateMethod;
+
+export function setIdentifierGenerationMethod(method: GenerationMethod): void {
+  configuredGenerationMethod = method;
+}
+
+export function setIdentifierUpdateMethod(method: UpdateMethod): void {
+  configuredUpdateMethod = method;
+}
+
+export function setIdentifierForgetMethod(method: ForgetMethod): void {
+  configuredForgetMethod = method;
+}
+
+export function setIdentifierResetMethod(method: ResetMethod): void {
+  configuredResetMethod = method;
+}
+
+function defaultGenerationMethod(data: ResourceIdentifierObject, bucket: string): string {
+  if (isNonEmptyString(data.lid)) {
+    return data.lid;
+  }
+  let { type, id } = data;
+  if (isNonEmptyString(id)) {
+    return `@ember-data:lid-${normalizeModelName(type)}-${id}`;
+  }
+  return uuidv4();
+}
+
+const IdentifierCaches = new WeakMap<object, IdentifierCache>();
+
+export function identifierCacheFor(store: object): IdentifierCache {
+  let cache = IdentifierCaches.get(store);
+
+  if (cache === undefined) {
+    cache = new IdentifierCache();
+    IdentifierCaches.set(store, cache);
+  }
+
+  return cache;
+}
+
+function defaultEmptyCallback(...args: any[]): any {}
+
+let DEBUG_MAP;
+if (DEBUG) {
+  DEBUG_MAP = new WeakMap<StableRecordIdentifier, StableRecordIdentifier>();
+}
+
+export class IdentifierCache {
+  // Typescript still leaks private properties in the final
+  // compiled class, so we may want to move these from _underscore
+  // to a WeakMap to avoid leaking
+  private _cache = {
+    lids: Object.create(null) as IdentifierMap,
+    types: Object.create(null) as TypeMap,
+  };
+  private _generate: GenerationMethod;
+  private _update: UpdateMethod;
+  private _forget: ForgetMethod;
+  private _reset: ResetMethod;
+
+  constructor() {
+    // we cache the user configuredGenerationMethod at init because it must
+    // be configured prior and is not allowed to be changed
+    this._generate = configuredGenerationMethod || defaultGenerationMethod;
+    this._update = configuredUpdateMethod || defaultEmptyCallback;
+    this._forget = configuredForgetMethod || defaultEmptyCallback;
+    this._reset = configuredResetMethod || defaultEmptyCallback;
+  }
+
+  // allows us to peek without generating when needed
+  // useful for the "create" case when we need to see if
+  // we are accidentally overwritting something
+  peekRecordIdentifier(resource: ResourceIdentifierObject, shouldGenerate: true): StableRecordIdentifier;
+  peekRecordIdentifier(resource: ResourceIdentifierObject, shouldGenerate: false): StableRecordIdentifier | undefined;
```

Comment:
In other places we use `peek` to mean look up but don't generate. Maybe we should keep that distinction here?

Reply:
I'll give this some thought